sys/key_rev_latest_value.js

```diff
@@ -0,0 +1,144 @@
+"use strict";
+
+var zlib = require('zlib');
+var P = require('bluebird');
+var uuid = require('cassandra-uuid').TimeUuid;
+var preq = require('preq');
+
+var HyperSwitch = require('hyperswitch');
+var URI = HyperSwitch.URI;
+
+var spec = HyperSwitch.utils.loadSpec(__dirname + '/key_rev_value.yaml');
+
+function ArchivalBucket(options) {
+}
+
+ArchivalBucket.prototype._latestName = function(bucket) {
+    return bucket + '.latest';
+};
+
+ArchivalBucket.prototype._archiveName = function(bucket) {
+    return bucket;
+};
+
+ArchivalBucket.prototype.createBucket = function(hyper, req) {
+    var self = this;
+    var rp = req.params;
+    var latestConfig = Object.assign({}, req.body, {
+        revisionRetentionPolicy: { type: 'latest_hash' }
+    });
+    if (latestConfig.valueType !== 'json') {
+        latestConfig.options = latestConfig.options || {};
+        latestConfig.options.compression = [];
+    }
+    return P.join(
+        hyper.put({
+            uri: new URI([rp.domain, 'sys', 'key_rev_value', self._latestName(rp.bucket)]),
+            headers: req.headers,
+            body: latestConfig
+        }),
+        hyper.put({
+            uri: new URI([rp.domain, 'sys', 'key_rev_value', self._archiveName(rp.bucket)]),
+            headers: req.headers,
+            body: req.body
+        })
+    )
+    .then(function() {
+        return { status: 201 }; });
+};
+
+function requestURI(rp, bucket) {
+    var requestPath = [rp.domain, 'sys', 'key_rev_value', bucket, rp.key];
+    if (rp.revision) {
+        requestPath.push('' + rp.revision);
+        if (rp.tid) {
+            requestPath.push(rp.tid);
+        }
+    }
+    return new URI(requestPath);
+}
+
+ArchivalBucket.prototype.getRevision = function(hyper, req) {
+    var self = this;
+    var rp = req.params;
+    return hyper.get({
+        uri: requestURI(rp, self._latestName(rp.bucket)),
+        headers: req.headers
+    })
+    .then(function(res) {
+        if (!/^application\/json/.test(res.headers['content-type'])) {
+            res.headers = Object.assign(res.headers, { 'content-encoding': 'gzip' });
+        }
+        return res;
+    })
+    .catch({ status: 404 }, function() {
+        return hyper.get({
+            uri: requestURI(rp, self._archiveName(rp.bucket)),
+            headers: req.headers
+        });
+    });
+};
+
+ArchivalBucket.prototype.listRevisions = function(hyper, req) {
+    var self = this;
+    var rp = req.params;
+    return hyper.get({
+        uri: new URI([rp.domain, 'sys', 'key_rev_value', self._archiveName(rp.bucket), rp.key, '']),
+        query: req.query
+    });
+};
+
+ArchivalBucket.prototype.putRevision = function(hyper, req) {
+    var self = this;
+    var rp = req.params;
+    var prepare;
+    if (/^application\/json/.test(req.headers['content-type'])) {
```

Comment:
We should really store JSON in compressed form as well. The main question is how to avoid breaking existing consumers expecting decoded JSON. For now, we can decompress & decode manually here. Longer term, we should figure out a way to defer that until it's really needed, for example by moving towards a `fetch` like response API with a `json()` method.